### PR TITLE
Configure teacher tasks manager

### PR DESCRIPTION
**Pull request by nbey**
TO DO:
- Configure Grid
  - [x] Implement Aggregrid
- Configure User Actions
  - [x] Create Task
  - [x] Rate Student Task
  - [x] Assign Student Task
  - [x] Edit Student Task
  - [x] Unassign Student Task

### CODE REVIEW

File: sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js [hunk 1]
```diff
@@ -1,43 +1,457 @@
...
+        },
+        'slate-tasks-teacher-appheader button[action=create]': {
+            click: 'onCreateTaskClick'
+        }
+    },
+
+    routes: {
+        'section/:sectionId': {
+            sectionId: '([a-zA-Z0-9])+',
+            action: 'viewCourseSection'
```
Comment (themightychris) [inline sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js]: @nbey our convention for naming route handles in `showCourseSection(sectionID)`. Please see https://github.com/JarvusInnovations/jarvus-guide-to-extjs/blob/master/applications/controllers.md

File: sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js [hunk 2]
```diff
@@ -1,43 +1,457 @@
...
+            datepicker.show().showAt(coords[0], coords[1] - datepicker.getHeight()); // subtract datepicker height from y coord to show above button.
+        }
+    },
+
+    onUnassignStudentTaskClick: function() {
+        var me = this,
+            taskRater = me.getTaskRater(),
+            studentTask = taskRater.getStudentTask();
+
+        Ext.Msg.confirm('Are you sure?', 'Do you want to unassign this student task? This can not be undone.', function(ans) {
```
Comment (themightychris) [inline sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js:284]: @nbey `Ext.MessageBox` needs to be explicitly listed in this class's `requires` if you're going to use the `Ext.Msg` singleton it provides

File: sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js [hunk 3]
```diff
@@ -1,43 +1,457 @@
...
+        var me = this,
+            coords = btn.getXY(),
+            datepicker;
+
+        if (btn.dateSelected) {
+            me.assignStudentTaskRevision(btn.dateSelected);
+        } else {
+            datepicker = Ext.widget({
+                xtype: 'datepicker',
+                floating: true,
```
Comment (themightychris) [inline sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js:268]: @nbey this controller needs to explicitly depend on the datepicker component via its `requires` array.

Also, you can use the ref system to cleanly define a "factory" for these pickers:

``` javascript
refs: {
    datePicker: {
        forceCreate: true,

        xtype: 'datepicker',
        floating: true
    }
}
```

(I'm not sure if a `selector` is still required, it would be useless with `forceCreate` but give the framework what it wants if it complains about it being missing)

and then you can do:

```
me.getDatePicker({
    handler: function(picker, date) {
        // ...
    }
}).showAt(...);
```

Comment (themightychris) [inline sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js:268]: @nbey Also, don't call show() and showBy separately, showBy handles showing and does it with some finer control to ensure it doesn't flash in the wrong position first.

File: sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js [hunk 4]
```diff
@@ -1,43 +1,457 @@
...
+
+        if (studentTask) {
+            taskStatus = studentTask.get('TaskStatus');
+            if (taskStatus === 'assigned') {
+                return me.editStudentTask(studentTask);
+            } else if (taskStatus === 'submitted' || taskStatus === 're-submitted' || taskStatus === 're-assigned') {
+                return me.rateStudentTask(studentTask);
+            } else if (taskStatus === 'completed') {
+                return me.rateStudentTask(studentTask, true);
+            }
```
Comment (themightychris) [inline sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js]: @nbey don't use `else` following `if` blocks that end in `return`, just do:

```
if (...) {
    return ...;
}

if (...) {
    return ...;
}
```

File: sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js [hunk 5]
```diff
@@ -1,43 +1,457 @@
...
+        return me.assignStudentTask(taskId, studentId);
+    },
+
+    onRateSkillClick: function(ratingView, ratingObject) {
+        var me = this,
+            taskRater = me.getTaskRater(),
+            studentTask = taskRater.getStudentTask();
+
+        Slate.API.request({
+            url: '/cbl/student-tasks/'+studentTask.getId()+'/rate',
```
Comment (themightychris) [inline sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js]: @nbey the `Ext.data.Model` class has a standard `toUrl()` method you can override and implement. Centralize this logic on the model so the code here can just be: `url: studentTask.toUrl() + '/rate'`

File: sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js [hunk 6]
```diff
@@ -1,43 +1,457 @@
...
+        taskEditor.setStudentTask(studentTask);
+        taskEditor.show();
+    },
+
+    editTask: function(taskRecord) {
+        var me = this,
+            taskEditor = me.getTaskEditor();
+
+        if (!taskRecord) {
+            taskRecord = Ext.create('Slate.cbl.model.Task');
```
Comment (themightychris) [inline sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js]: @nbey this is an implicit dependency on the model, avoid calling `Ext.create` or `Ext.widget` within a controller.

Put at the top of the controller:

```
models: [
    'Task@Slate.cbl.model'
]
```

and then you can do: `me.getTaskModel().create({ ...})`

File: sencha-workspace/SlateTasksTeacher/app/store/StudentTasks.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
+Ext.define('SlateTasksTeacher.store.StudentTasks', {
+    extend: 'Ext.data.Store',
+    requires: [
+        'Slate.cbl.model.StudentTask'
```
Comment (themightychris) [inline sencha-workspace/SlateTasksTeacher/app/store/StudentTasks.js]: @nbey this is redundant, stores add their `model` to the requires list automatically.

File: sencha-workspace/SlateTasksTeacher/app/store/StudentTasks.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    ],
+
+    model: 'Slate.cbl.model.StudentTask',
+    remoteFilter: true,
+    remoteSort: true,
+    pageSize: 0,
+
+    proxy: {
+        type: 'slate-records',
```
Comment (themightychris) [inline sencha-workspace/SlateTasksTeacher/app/store/StudentTasks.js:13]: @nbey this class needs to require the proxy class too if it's going to require this `type`. The model could get reconfigured one day to use a different proxy

File: sencha-workspace/SlateTasksTeacher/app/store/Students.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
+Ext.define('SlateTasksTeacher.store.Students',{
+    extend: 'Slate.cbl.store.Students',
+    requires: [
+        'Emergence.ext.proxy.Records'
+    ],
+    proxy: {
```
Comment (themightychris) [inline sencha-workspace/SlateTasksTeacher/app/store/Students.js]: @nbey the proxy config should be down below `config`, not up with the `Ext.define` options

File: sencha-workspace/SlateTasksTeacher/app/view/AppHeader.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        align: 'left'
+    },
+
+    items: [{
+        xtype: 'component',
+        cls: 'slate-appheader-title',
+        itemId: 'title',
+        html: 'Teacher Task Manager'
+    },{
+        xtype: 'combo',
```
Comment (themightychris) [inline sencha-workspace/SlateTasksTeacher/app/view/AppHeader.js:21]: @nbey combo needs to be required

File: sencha-workspace/SlateTasksTeacher/app/view/Dashboard.js [hunk 11]
```diff
@@ -23,6 +28,10 @@ Ext.define('SlateTasksTeacher.view.Dashboard', {
         return Ext.factory(gridLegend, 'SlateTasksTeacher.view.GridLegend', oldGridLegend);
     },
 
+    updateCourseSection: function(courseSection) {
+        this.fireEvent('coursesectionselect', courseSection);
```
Comment (themightychris) [inline sencha-workspace/SlateTasksTeacher/app/view/Dashboard.js]: @nbey the first parameter to any component event should be a reference to the component firing the event (`this` in this case)

File: sencha-workspace/SlateTasksTeacher/app/view/StudentsGrid.js [hunk 12]
```diff
@@ -2,312 +2,233 @@
...
+                    }
+
+                    return html;
+                }
             }
-        ]
-    },
+        ],
+
+        tpl: [
```
Comment (themightychris) [inline sencha-workspace/SlateTasksTeacher/app/view/StudentsGrid.js:95]: @nbey why are all these structural tpls' being overridden in this implementation? They're not supposed to be customized and overriding them here is dangerous

File: sencha-workspace/SlateTasksTeacher/app/view/StudentsGrid.js [hunk 13]
```diff
@@ -2,312 +2,233 @@
...
-        scope: 'this',
-        click: {
-            fn: 'onGridClick',
-            element: 'el',
-            delegate: '.slate-studentsgrid-row'
-        }
+        me.callParent(arguments);
+        me.setConfig({
+            cellRenderer: me.cellRenderFn,
+            subCellRenderer: me.cellRenderFn
```
Comment (themightychris) [inline sencha-workspace/SlateTasksTeacher/app/view/StudentsGrid.js:214]: @nbey why is this necessary? The function bodies should just be declared directly in the config section

File: sencha-workspace/SlateTasksTeacher/app/view/TaskEditor.js [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+            assignmentsfieldCheckbox = assignmentsfield.down('checkboxfield'),
+            titleField = me.down('slate-tasks-titlefield'),
+            parentTaskField = titleField.next('slate-tasks-titlefield'),
+            experienceField = me.down('#experience-type'),
+            duedateField = me.down('#duedate'),
+            expirationdateField = me.down('#expirationdate'),
+            skillsField = me.down('slate-skillsfield'),
+            attachmentsfield = me.down('slate-tasks-attachmentsfield'),
+            instructionsField = me.down('#instructions'),
+            taskPrivacyField = me.down('#status');
```
Comment (themightychris) [inline sencha-workspace/SlateTasksTeacher/app/view/TaskEditor.js:23]: @nbey consider providing a `getForm()` method on the modal, grabbing that once, and then using form`s`getField` api for this to grab form field based on their name as intended rather than with itemIds

File: sencha-workspace/packages/slate-cbl/src/model/StudentTask.js [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+    proxy: {
+        type: 'slate-records',
+        url: '/cbl/student-tasks'
+    },
+
+    getTaskSkillsGroupedByCompetency: function() {
+        var comps = [], compIds = [],
+            skills = this.get('TaskSkills');
+
+        Ext.each(skills, function(skill) {
```
Comment (themightychris) [inline sencha-workspace/packages/slate-cbl/src/model/StudentTask.js]: @nbey use for loops instead of `Ext.each`

File: sencha-workspace/packages/slate-cbl/src/view/AttachmentsList.js [hunk 16]
```diff
@@ -13,20 +13,29 @@ Ext.define('Slate.cbl.view.AttachmentsList', {
...
+                '<li class="slate-attachmentslist-item <tpl if="kind">slate-attachment-{kind}</tpl>">',
+                    '<span class="slate-attachment-title"><a href="{URL}" target=_blank>{title}</a></span>',
+                    '<tpl if="this.owner.getEditable()">',
+                        '<button class="plain" action="settings"><i class="fa fa-gear"></i></button>',
+                        '<button class="plain" action="remove"><i class="fa fa-times-circle"></i></button>',
+                    '</tpl>',
+                '</li>',
+            '</tpl>'
+        ];
+        me.callParent(arguments);
```
Comment (themightychris) [inline sencha-workspace/packages/slate-cbl/src/view/AttachmentsList.js]: @nbey you shouldn't be calling `initConfig`, that already gets called in the component's lifecycle. Why was it necessary to set tpl from initComponent instead of just overriding the property? An initComponent override looking like this shouldn't exist

File: sencha-workspace/packages/slate-cbl/src/widget/AttachmentsField.js [hunk 17]
```diff
@@ -42,6 +46,29 @@ Ext.define('Slate.cbl.widget.AttachmentsField', {
         disabled: true
     }],
 
+    updateReadOnly: function(readOnly) {
+        if (!this.rendered) {
+            return this.on('render', function() {
+                return this.updateReadOnly(readOnly);
+            });
```
Comment (themightychris) [inline sencha-workspace/packages/slate-cbl/src/widget/AttachmentsField.js]: @nbey be sure to pass the `{ single: true }` option with this `.on()` call, or this component being rendered within something like a tab panel or window that can be rendered/unrendered several times will result in an infinitely growing stack of this function getting called for each update

File: sencha-workspace/packages/slate-cbl/src/widget/RatingView.js [hunk 18]
```diff
@@ -74,5 +45,102 @@ Ext.define('Slate.cbl.widget.RatingView', {
...
+            }];
+        if (!me.getReadOnly()) {
+            if (!menu && menuRatings.length) {
+                Ext.each(menuRatings, function(mr) {
+                    items.push({
+                        text: mr,
+                        value: mr
+                    });
+                });
+                me.setMenu(menu = Ext.create('Ext.menu.Menu', {
```
Comment (themightychris) [inline sencha-workspace/packages/slate-cbl/src/widget/RatingView.js:127]: @nbey this class should explicitly require `Ext.menu.Menu`

File: sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js (hunk 3 above)
Comment (nbey) [inline sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js:268]: @themightychris I'm going to remove this code during my next work session (we are no longer going to use the datepicker like this)

File: sencha-workspace/SlateTasksTeacher/app/view/StudentsGrid.js (hunk 12 above)
Comment (nbey) [inline sencha-workspace/SlateTasksTeacher/app/view/StudentsGrid.js:95]: @themightychris I changed the name of the field used in the template, `{id}`, to match the models id field, `{ID}`. Should I create a non-persisted field instead?

File: sencha-workspace/SlateTasksTeacher/app/view/StudentsGrid.js (hunk 13 above)
Comment (nbey) [inline sencha-workspace/SlateTasksTeacher/app/view/StudentsGrid.js:214]: @themightychris Not necessary, just thought it would be better for them to use the same exact render function without duplicating the code.

File: sencha-workspace/packages/slate-cbl/src/view/AttachmentsList.js (hunk 16 above)
Comment (nbey) [inline sencha-workspace/packages/slate-cbl/src/view/AttachmentsList.js]: @themightychris I was having trouble with the configuring the template -- It seems that the template object doesn't have an updated reference to the container it's displayed in.
`this.owner.getEditable()` appears to _only_ return it's **initial value**. Any suggestions?

File: sencha-workspace/SlateTasksTeacher/app/view/StudentsGrid.js (hunk 13 above)
Comment (themightychris) [reply sencha-workspace/SlateTasksTeacher/app/view/StudentsGrid.js:214]: @nbey better approach (doesn't mess with initComponent lifecycle): 4170cf3f80a7ff81b87717f468b191d6bd6073ac

File: sencha-workspace/SlateTasksTeacher/app/view/StudentsGrid.js (hunk 12 above)
Comment (themightychris) [reply sencha-workspace/SlateTasksTeacher/app/view/StudentsGrid.js:95]: The aggregrid should be adoptable to any store you want to provide, I just resolved the need for this in aggregrid's master branch
